tensorflow/compiler/xla/service/gpu/gpu_hlo_schedule.cc

```diff
@@ -184,6 +185,86 @@ void BFSLaunchOrder(const HloComputation* computation,
   }
 }
 
+bool CustomCallWithSchedule(const HloInstruction* instr,
+                            CustomCallSchedule schedule) {
+  return instr->opcode() == HloOpcode::kCustomCall &&
+         static_cast<const HloCustomCallInstruction*>(instr)
+                 ->custom_call_schedule() == schedule;
+}
+
+// Schedules EARLIEST and LATEST custom-calls. This supports a custom-call use
+// case, where a logical operation is lowered into two HLOs (e.g., PerformX and
+// PerformXDone). We utilize this mechanism to either hide host latencies
+// between the pair of the custom-calls or more accurately identify the def-use
+// relationship of the two calls (typically PerformX is scheduled right after
+// all of its producers have been scheduled and PerformXDone is scheduled right
+// before its first consumer.)
+HloInstructionSequence postprocessor_to_custom_schedule(
+    const HloInstructionSequence& input) {
+  // Schedule `EARLIEST`.
+  std::deque<HloInstruction*> earliest_scheduled;
+  {
+    absl::flat_hash_set<HloInstruction*> scheduled;
+    auto is_scheduled = [&](const HloInstruction* instr) -> bool {
+      return scheduled.contains(instr);
+    };
+    const std::vector<HloInstruction*>& instrs = input.instructions();
+    for (HloInstruction* instr : instrs) {
+      if (scheduled.contains(instr)) {
```

Comment:
Haven't we just defined the lambda checking precisely this? Or more concretely: do we need the lambda at all then?

Reply:
The lambda is needed because the absl::c_all_of() below requires a function form.

`            absl::c_all_of(user->operands(), is_scheduled) &&
            absl::c_all_of(user->control_predecessors(), is_scheduled)) {`

It is cleaner to also use the lambda here. Will update it.

Reply:
Updated.